sklearn/preprocessing/_function_transformer.py

```diff
@@ -108,9 +109,15 @@ def _check_input(self, X):
 
     def _check_inverse_transform(self, X):
         """Check that func and inverse_func are the inverse."""
+        X = np.asarray(X) if not hasattr(X, "dtype") else X
         idx_selected = slice(None, None, max(1, X.shape[0] // 100))
         X_round_trip = self.inverse_transform(self.transform(X[idx_selected]))
-        if not _allclose_dense_sparse(X[idx_selected], X_round_trip):
+
+        if np.issubdtype(X.dtype, np.number):
+            valid = _allclose_dense_sparse(X[idx_selected], X_round_trip)
+        else:
+            valid = np.all(X == X_round_trip)
```

Comment:
If `inverse_func` returns a dataframe of strings & floats, this can fail for floats:

```python
import pandas as pd
import numpy as np

X = pd.DataFrame({"a": ["first"], "a_int": [1], "a_float": [0.3]})
X_round_trip = pd.DataFrame({"a": ["first"], "a_int": [1], "a_float": [0.1 + 0.1 + 0.1]})

X_np = np.asarray(X)
np.all(X_np == X_round_trip)  # False
```

I would prefer not to deal with mixed types when checking for the inverse in `FunctionTransformer`. I think we can error when `X_round_trip` is a not a numerical ndarray and suggest `check_inverse=False`.

WDYT @glemaitre ?

Reply:
This is probably the best. I would have like to support when we have only strings but it would be difficult to detect. Only the ndarray with `"S"` or `"U"` dtype could be easier to handle.

Reply:
This still needs to be address. I think we should error when `X_round_trip` is a dataframe that is not all numerical and suggest `check_inverse=False`.

Checking the inverse with DataFrames with all numerical dtypes works on `main`, so we need to be careful to continue to support that use case.

Reply:
I've updated `_check_inverse_transform` as suggested, now it raises error when input is not all numerical